libsolidity/analysis/ControlFlowRevertPruner.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/ControlFlowRevertPruner.h>
+
+#include <libsolutil/Algorithms.h>
+
+namespace solidity::frontend
+{
+
+void ControlFlowRevertPruner::run(ASTNode const& _astRoot)
+{
+	// Run first iteration without following recursive loops
+	_astRoot.accept(*this);
+
+	// In the second iteration we process anything left pending
+	removePendingPaths();
+}
+
+void ControlFlowRevertPruner::removePendingPaths()
+{
+	// Anything still left pending must be a infinite recursive loop.
+	for (auto& [key, val]: m_functionReverts)
+	{
+		RevertState& revertState = std::get<State>(val);
+		auto functionFlow = m_cfg.functionFlow(
+			*std::get<Function>(key),
+			std::get<Contract>(key)
+		);
+
+		if (CFGNode* node = std::get<Node>(val))
+			for (auto const* functionCall: node->functionCalls)
+				switch (checkForReverts(*functionCall, std::get<Contract>(key)))
+				{
+				// Remove exits & add revert node
+				case RevertState::AllPathsRevert:
+					[[fallthrough]];
+				case RevertState::Pending:
+					node->exits.clear();
+					node->exits.push_back(functionFlow.revert);
+					break;
+				case RevertState::HasNonRevertingPath:
+					break;
+				}
+
+		if (revertState == RevertState::Pending)
+			revertState = RevertState::AllPathsRevert;
+	}
+}
+
+void ControlFlowRevertPruner::removeRevertingPaths(FunctionDefinition const& _function, ContractDefinition const* _contract)
+{
+	auto& functionFlow = m_cfg.functionFlow(_function, _contract);
+	std::tuple const revertMapKey = {_contract, &_function};
+
+	std::get<State>(m_functionReverts[revertMapKey]) = RevertState::Pending;
+
+	// Traverse the function flow and remove exits from nodes with reverting
+	// function calls.
+	// Recursive loops are saved for later processing.
+	traverseFunctionFlow(_function, _contract, functionFlow,
+		[&](RevertState _state, CFGNode* _node, FunctionFlow const& _flow) {
+			if (_state == RevertState::AllPathsRevert)
+			{
+				// Remove exits if call always reverts
+				_node->exits.clear();
+				_node->exits.push_back(_flow.revert);
+			}
+			else if (_state == RevertState::Pending)
+			{
+				// Remember node for later re-evaluation
+				std::get<Node>(m_functionReverts[revertMapKey]) = _node;
```

Comment:
I don't understand why this suffices, but maybe at the end of the review.

Reply:
It will be processed by `removePendingPaths()` in a later step